### PR TITLE
Added singleMarkable option and get-marked event

### PR DESCRIPTION
**Pull request by ReneZumtobel**
| Q | A
| --- | ---
| Bug fix? | yes
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | https://github.com/sulu/sulu/pull/2876
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

Adds singleMarkable option and get-marked event.

### CODE REVIEW

File: husky_components/column-navigation/main.js [hunk 1]
```diff
@@ -1281,13 +1299,46 @@ define(function() {
...
+                if (this.options.singleMarkable === true) {
+                    $.each(this.marked, function (key, value) {
+                        this.unmark(value);
+                    }.bind(this));
+                }
+                this.sandbox.dom.addClass($element, constants.markedClass);
+                this.marked.push(id);
+                this.setItemsTextWidth($element);
+                // change unmark to mark icon
+                var $markButton = $element.find('span:last-child > .'+ this.options.actionIcon);
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: Space around `+`

File: husky_components/column-navigation/main.js [hunk 2]
```diff
@@ -1281,13 +1299,46 @@ define(function() {
...
+                $markButton.removeClass(this.options.actionIcon);
+
+                /* prepend icon class */
+                var classes = $markButton.attr('class');
+                classes = this.options.unmarkIcon + ' ' + classes;
+                $markButton.attr('class', classes);
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: Any reason this has to be done that complicated? Why not simply using `addClass`?

File: husky_components/column-navigation/main.js [hunk 3]
```diff
@@ -1281,13 +1299,46 @@ define(function() {
...
             var $element = this.$find('li[data-id="' + id + '"]');
             if (!!$element.length) {
                 this.sandbox.dom.removeClass($element, constants.markedClass);
+                // change unmark to mark icon
+                var $unmarkButton = $element.find('span:last-child > .'+ this.options.unmarkIcon);
+                $unmarkButton.removeClass(this.options.unmarkIcon);
+                /* prepend icon class */
+                var classes = $unmarkButton.attr('class');
+                classes = this.options.actionIcon + ' ' + classes;
+                $unmarkButton.attr('class', classes);
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: Same as above

File: husky_components/column-navigation/main.js [hunk 4]
```diff
@@ -1576,9 +1627,11 @@ define(function() {
             item = this.columns[column][id];
 
             if (this.options.markable === true) {
-                this.sandbox.dom.addClass($listItem, constants.markedClass);
-                this.marked.push(id);
-                this.setItemsTextWidth($listItem);
+                if ($.inArray(id,this.marked) < 0) {
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: space after `,`

File: husky_components/column-navigation/main.js [hunk 5]
```diff
@@ -1587,6 +1640,18 @@ define(function() {
...
                 this.sandbox.emit(ACTION.call(this), item);
             }
+        },
+
+        /**
+         * Returns all the marked objects.
+         */
+        getMarkedHandler: function() {
+            var elements = [];
+            $.each(this.marked, function (key, markedElement) {
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: I would say the [map function](https://developer.mozilla.org/de/docs/Web/JavaScript/Reference/Global_Objects/Array/map) is a better match here.

File: husky_components/column-navigation/main.js (hunk 2 above)
Comment (ReneZumtobel) [inline husky_components/column-navigation/main.js]: First I tried using the `addClass` Function, but that method didn´t work, that´s why I used this way.

Comment (danrot) [inline husky_components/column-navigation/main.js]: Why didn't it work?

Comment (ReneZumtobel) [inline husky_components/column-navigation/main.js]: It added the class but the icon wouldn´t show. But when I copied the class as the first class it worked.

File: husky_components/column-navigation/main.js [hunk 6]
```diff
@@ -804,6 +817,10 @@ define(function() {
             var $container = this.sandbox.dom.find('.' + constants.iconsRightClass, $item),
                 actionIcon = getActionIcon.call(this, data);
 
+            if (this.marked.indexOf(data[this.options.idName]) !== -1) { // if is marked change action Icon
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: Move the comment into an own line

File: husky_components/column-navigation/main.js [hunk 7]
```diff
@@ -1281,14 +1303,44 @@ define(function() {
...
+            var $element = this.$find('li[data-id="' + id + '"]');
+            if (!!$element.length) {
+                // if single markable unmark all the others
+                if (this.options.singleMarkable === true) {
+                    $.each(this.marked, function (key, value) {
+                        this.unmark(value);
+                    }.bind(this));
+                }
+                this.sandbox.dom.addClass($element, constants.markedClass);
+                if($.inArray(id,this.marked) < 0) {
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: space after `,`

File: husky_components/column-navigation/main.js [hunk 8]
```diff
@@ -1281,14 +1303,44 @@ define(function() {
...
          * Unmarks a node for a given id
          * @param id {Number|String} the id of the node to unmark
          */
         unmark: function(id) {
             var $element = this.$find('li[data-id="' + id + '"]');
             if (!!$element.length) {
-                this.sandbox.dom.removeClass($element, constants.markedClass);
-                this.marked.splice(this.marked.indexOf(id), 1);
+                $element.removeClass(constants.markedClass);
+                if($.inArray(id,this.marked) >= 0) {
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: space after `,`

File: husky_components/column-navigation/main.js [hunk 9]
```diff
@@ -1281,14 +1303,44 @@ define(function() {
...
+                    this.marked.splice($.inArray(id, this.marked), 1);
+                }
+                // change unmark to mark icon
+                var $unmarkButton = $element.find('span:last-child > .'+ this.options.unmarkIcon);
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: spaces around `+`

File: husky_components/column-navigation/main.js [hunk 10]
```diff
@@ -64,6 +66,7 @@ define(function() {
             instanceName: '',
             hasSubName: 'hasSub',
             actionIcon: 'fa-pencil',
+            unmarkIcon: 'fa-minus',
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: we could use `fa-minus-circle` here

File: husky_components/column-navigation/main.js [hunk 11]
```diff
@@ -446,8 +459,15 @@ define(function() {
 
             this.columns = [];
             this.selected = [];
-            // array with all marked ids
-            this.marked = this.options.premarkedIds || [];
+
+            this.marked = {};
+            // write premerkedIds in this.marked
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: `premarkedIds`, or should we delete this comment completely? Doesn't feel like it adds a huge value.

File: husky_components/column-navigation/main.js [hunk 12]
```diff
@@ -446,8 +459,15 @@ define(function() {
...
+            if (this.options.premarkedIds) {
+                this.options.premarkedIds.forEach(function (premarkedId) {
+                    // write empry objekt as value because element not available jet
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: `empty object` and `yet`

File: husky_components/column-navigation/main.js [hunk 13]
```diff
@@ -446,8 +459,15 @@ define(function() {
...
+                    this.marked[premarkedId] = {};
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: Will this be written correctly later for all the objects?

Comment (ReneZumtobel) [inline husky_components/column-navigation/main.js]: Yes see line 695 i even added a comment

File: husky_components/column-navigation/main.js [hunk 14]
```diff
@@ -446,8 +459,14 @@ define(function() {
 
             this.columns = [];
             this.selected = [];
-            // array with all marked ids
-            this.marked = this.options.premarkedIds || [];
+
+            this.marked = {};
+            if (this.options.premarkedIds) {
+                this.options.premarkedIds.forEach(function (premarkedId) {
+                    // write empty object as value because element not available jet
```
Comment (danrot) [inline husky_components/column-navigation/main.js]: `yet`

**Pull request merged**